File: frontend/src/utils/fileFormat.ts
// frontend/src/utils/fileFormat.ts

// Mirror of backend/app/utils/file_format.py — keep in sync.
export const ALLOWED_UPLOAD_EXTENSIONS = ['.csv', '.xlsx'] as const;

export const ALLOWED_UPLOAD_ACCEPT =
  '.csv,.xlsx,text/csv,application/vnd.openxmlformats-officedocument.spreadsheetml.sheet';

export function extensionOf(filename: string): string {
  const dot = filename.lastIndexOf('.');
  if (dot <= 0 || dot === filename.length - 1) return '';
  return filename.slice(dot).toLowerCase();
}

export function isAllowedUpload(filename: string): boolean {
  return (ALLOWED_UPLOAD_EXTENSIONS as readonly string[]).includes(extensionOf(filename));
}

File: backend/app/utils/file_format.py
"""Single source of truth for accepted upload file formats.

Used by the upload router, the detect-headers helper, and the tabular parser
so the allow-list is declared exactly once. The frontend mirrors this list in
src/utils/fileFormat.ts — keep them in sync.
"""

from __future__ import annotations

ALLOWED_UPLOAD_EXTENSIONS: frozenset[str] = frozenset({".csv", ".xlsx"})


def extension_of(filename: str) -> str:
    """Return the lower-cased extension including the dot, or '' if absent."""
    prefix, dot, ext = filename.rpartition(".")
    if not dot or not ext or not prefix:
        return ""
    return f".{ext.lower()}"


def is_allowed_upload(filename: str) -> bool:
    return extension_of(filename) in ALLOWED_UPLOAD_EXTENSIONS
